src/complete.rs

```diff
@@ -1,82 +1,160 @@
+//! ABNF Core Rules (RFC5234 B.1.)
+
+use std::ops::{RangeFrom, RangeTo};
+
 use nom::{
-    branch::alt,
-    bytes::complete::{tag, take},
-    character::complete::{self as character, anychar, line_ending},
-    combinator::{opt, recognize, verify},
-    error::ParseError,
+    character::complete::satisfy,
+    combinator::{opt, recognize},
+    error::{ErrorKind, ParseError},
     multi::many0_count,
-    sequence::terminated,
-    IResult,
+    sequence::{pair, terminated},
+    AsChar, Err as OutCome, IResult, InputIter, InputLength, Offset, Slice,
 };
 
-use crate::*;
+use crate::{
+    is_alpha, is_bit, is_char, is_cr, is_ctl, is_digit, is_dquote, is_hexdig, is_htab, is_lf,
+    is_sp, is_wsp,
+};
 
 /// ALPHA = %x41-5A / %x61-7A ; A-Z / a-z
-pub fn alpha<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, char, E> {
-    verify(anychar, |&c| is_alpha(c))(input)
+pub fn alpha<I, E>(input: I) -> IResult<I, char, E>
+where
+    I: InputIter + Slice<RangeFrom<usize>>,
+    <I as InputIter>::Item: AsChar,
+    E: ParseError<I>,
+{
+    satisfy(is_alpha)(input)
 }
 
 /// BIT = "0" / "1"
-pub fn bit<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, char, E> {
-    verify(anychar, |&c| is_bit(c))(input)
+pub fn bit<I, E>(input: I) -> IResult<I, char, E>
+where
+    I: InputIter + Slice<RangeFrom<usize>>,
+    <I as InputIter>::Item: AsChar,
+    E: ParseError<I>,
+{
+    satisfy(is_bit)(input)
 }
 
 /// CHAR = %x01-7F ; any 7-bit US-ASCII character, excluding NUL
-pub fn char<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, char, E> {
-    verify(anychar, |&c| is_char(c))(input)
-}
-
-/// CR = %x0D ; carriage return
-pub fn cr<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, char, E> {
-    character::char('\r')(input)
+pub fn char<I, E>(input: I) -> IResult<I, char, E>
+where
+    I: InputIter + Slice<RangeFrom<usize>>,
+    <I as InputIter>::Item: AsChar,
+    E: ParseError<I>,
+{
+    satisfy(is_char)(input)
 }
 
-pub fn crlf_strict<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, &str, E> {
-    tag("\r\n")(input)
-}
-
-pub fn crlf_relaxed<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, &str, E> {
-    line_ending(input)
+/// Carriage return
+///
+/// CR = %x0D
+pub fn cr<I, E>(input: I) -> IResult<I, char, E>
+where
+    I: InputIter + Slice<RangeFrom<usize>>,
+    <I as InputIter>::Item: AsChar,
+    E: ParseError<I>,
+{
+    satisfy(is_cr)(input)
 }
 
-/// CRLF = CR LF ; Internet standard newline
+/// Internet standard newline
+///
+/// CRLF = CR LF
 ///
 /// Note: this variant will strictly expect "\r\n".
 /// Use [crlf_relaxed](fn.crlf_relaxed.html) to accept "\r\n" as well as only "\n".
-pub fn crlf<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, &str, E> {
-    crlf_strict(input)
+pub fn crlf<I, E>(input: I) -> IResult<I, (char, char), E>
```

Comment:
function signature change a little, I choice to not use `tag` cause it's was more limited with generic and so to avoid using `recognize` I return tuple of char. Could be better with a zero sized CrLf structure